File: engine/internal/grpcapi/orchestration.go
package grpcapi

import (
	"context"

	"github.com/google/uuid"
	pb "github.com/edicorai/vibetown/engine/internal/grpcapi/gen/vibetown/orchestration/v1"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/emptypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// HealthHandler implements the HealthService gRPC service.
type HealthHandler struct {
	pb.UnimplementedHealthServiceServer
}

// NewHealthHandler returns a new HealthHandler.
func NewHealthHandler() *HealthHandler {
	return &HealthHandler{}
}

// Check returns the serving status of the engine.
func (h *HealthHandler) Check(_ context.Context, _ *pb.HealthCheckRequest) (*pb.HealthCheckResponse, error) {
	return &pb.HealthCheckResponse{
		Status: pb.HealthCheckResponse_SERVING_STATUS_SERVING,
	}, nil
}

// OrchestrationHandler implements the OrchestrationService gRPC service.
// It delegates to an Engine for all operations.
type OrchestrationHandler struct {
	pb.UnimplementedOrchestrationServiceServer

	engine Engine
}

// NewOrchestrationHandler returns a new OrchestrationHandler backed by the
// given Engine. If engine is nil, a default InMemoryEngine is used.
func NewOrchestrationHandler(engine ...Engine) *OrchestrationHandler {
	var e Engine
	if len(engine) > 0 && engine[0] != nil {
		e = engine[0]
	} else {
		e = NewInMemoryEngine()
	}
	return &OrchestrationHandler{engine: e}
}

// Engine returns the underlying Engine (useful for testing and server wiring).
func (o *OrchestrationHandler) Engine() Engine {
	return o.engine
}

// --- Town lifecycle ---

func (o *OrchestrationHandler) CreateTown(_ context.Context, req *pb.CreateTownRequest) (*pb.Town, error) {
	now := timestamppb.Now()
	return &pb.Town{
		Id:        uuid.NewString(),
		Name:      req.GetName(),
		Owner:     req.GetOwner(),
		CreatedAt: now,
		UpdatedAt: now,
	}, nil
}

func (o *OrchestrationHandler) GetTown(_ context.Context, req *pb.GetTownRequest) (*pb.Town, error) {
	if req.GetId() == "" {
		return nil, status.Error(codes.InvalidArgument, "id is required")
	}
	info, err := o.engine.GetTownConfig()
	if err != nil {
		return nil, status.Errorf(codes.Internal, "getting town config: %v", err)
	}
	return &pb.Town{
		Id:        info.ID,
		Name:      info.Name,
		Owner:     info.Owner,
		CreatedAt: timestamppb.New(info.CreatedAt),
		UpdatedAt: timestamppb.Now(),
	}, nil
}

// --- Rig management ---

func (o *OrchestrationHandler) CreateRig(_ context.Context, req *pb.CreateRigRequest) (*pb.Rig, error) {
	now := timestamppb.Now()
	return &pb.Rig{
		Id:          uuid.NewString(),
		TownId:      req.GetTownId(),
		Name:        req.GetName(),
		RepoUrl:     req.GetRepoUrl(),
		BeadsPrefix: req.GetBeadsPrefix(),
		CreatedAt:   now,
		UpdatedAt:   now,
	}, nil
}

func (o *OrchestrationHandler) ListRigs(_ context.Context, _ *pb.ListRigsRequest) (*pb.ListRigsResponse, error) {
	rigs, err := o.engine.ListRigs()
	if err != nil {
		return nil, status.Errorf(codes.Internal, "listing rigs: %v", err)
	}
	var pbRigs []*pb.Rig
	for _, r := range rigs {
		pbRigs = append(pbRigs, &pb.Rig{
			Id:          r.Name,
			Name:        r.Name,
			RepoUrl:     r.GitURL,
			BeadsPrefix: r.Prefix,
		})
	}
	return &pb.ListRigsResponse{Rigs: pbRigs}, nil
}

// --- Agent lifecycle ---

func (o *OrchestrationHandler) SpawnAgent(_ context.Context, req *pb.SpawnAgentRequest) (*pb.Agent, error) {
	info, err := o.engine.SpawnAgent(
		req.GetName(),
		req.GetRole(),
		req.GetRigId(),
		req.GetRuntime(),
		req.GetConfigJson(),
	)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "spawning agent: %v", err)
	}
	return agentInfoToProto(info), nil
}

func (o *OrchestrationHandler) KillAgent(_ context.Context, req *pb.KillAgentRequest) (*emptypb.Empty, error) {
	if err := o.engine.KillAgent(req.GetId()); err != nil {
		return nil, status.Errorf(codes.NotFound, "agent %q not found", req.GetId())
	}
	return &emptypb.Empty{}, nil
}

func (o *OrchestrationHandler) GetAgentStatus(_ context.Context, req *pb.GetAgentStatusRequest) (*pb.Agent, error) {
	info, err := o.engine.GetAgent(req.GetId())
	if err != nil {
		return nil, status.Errorf(codes.NotFound, "agent %q not found", req.GetId())
	}
	return agentInfoToProto(info), nil
}

func (o *OrchestrationHandler) ListAgents(_ context.Context, req *pb.ListAgentsRequest) (*pb.ListAgentsResponse, error) {
	agents, err := o.engine.ListAgents(
		req.GetRigId(),
		req.GetRole(),
		req.GetStatus(),
	)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "listing agents: %v", err)
	}
	var pbAgents []*pb.Agent
	for _, a := range agents {
		pbAgents = append(pbAgents, agentInfoToProto(a))
	}
	return &pb.ListAgentsResponse{Agents: pbAgents}, nil
}

// --- Convoy operations ---

func (o *OrchestrationHandler) StartConvoy(_ context.Context, req *pb.StartConvoyRequest) (*pb.Convoy, error) {
	info, err := o.engine.StartConvoy(
		req.GetName(),
		req.GetFormula(),
		req.GetConfigJson(),
	)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "starting convoy: %v", err)
	}
	return convoyInfoToProto(info), nil
}

func (o *OrchestrationHandler) GetConvoyStatus(_ context.Context, req *pb.GetConvoyStatusRequest) (*pb.Convoy, error) {
	if req.GetId() == "" {
		return nil, status.Error(codes.InvalidArgument, "id is required")
	}
	info, err := o.engine.GetConvoy(req.GetId())
	if err != nil {
		return nil, status.Errorf(codes.NotFound, "convoy %q not found", req.GetId())
	}
	return convoyInfoToProto(info), nil
}

func (o *OrchestrationHandler) ListConvoys(_ context.Context, req *pb.ListConvoysRequest) (*pb.ListConvoysResponse, error) {
	convoys, err := o.engine.ListConvoys(req.GetStatus())
	if err != nil {
		return nil, status.Errorf(codes.Internal, "listing convoys: %v", err)
	}
	var pbConvoys []*pb.Convoy
	for _, c := range convoys {
		pbConvoys = append(pbConvoys, convoyInfoToProto(c))
	}
	return &pb.ListConvoysResponse{Convoys: pbConvoys}, nil
}

// --- Work dispatch ---

func (o *OrchestrationHandler) DispatchWork(_ context.Context, req *pb.DispatchWorkRequest) (*pb.DispatchWorkResponse, error) {
	return &pb.DispatchWorkResponse{
		WorkItemId: uuid.NewString(),
		Accepted:   true,
	}, nil
}

// --- Merge queue ---

func (o *OrchestrationHandler) QueueMerge(_ context.Context, req *pb.QueueMergeRequest) (*pb.MergeRequest, error) {
	info, err := o.engine.QueueMerge(
		req.GetRigId(),
		req.GetBranch(),
		req.GetTargetBranch(),
		req.GetWorkItemId(),
		req.GetAgentId(),
	)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "queueing merge: %v", err)
	}
	return mergeInfoToProto(info), nil
}

func (o *OrchestrationHandler) GetMergeQueue(_ context.Context, req *pb.GetMergeQueueRequest) (*pb.MergeQueueResponse, error) {
	items, err := o.engine.GetMergeQueue(req.GetRigId())
	if err != nil {
		return nil, status.Errorf(codes.Internal, "getting merge queue: %v", err)
	}
	var pbItems []*pb.MergeRequest
	for _, item := range items {
		pbItems = append(pbItems, mergeInfoToProto(item))
	}
	return &pb.MergeQueueResponse{Requests: pbItems}, nil
}

// --- Proto conversion helpers ---

func agentInfoToProto(a *AgentInfo) *pb.Agent {
	return &pb.Agent{
		Id:             a.ID,
		Name:           a.Name,
		Role:           a.Role,
		RigId:          a.RigID,
		Status:         a.Status,
		Runtime:        a.Runtime,
		ConfigJson:     a.ConfigJSON,
		LastActivityAt: timestamppb.New(a.LastActivityAt),
		CreatedAt:      timestamppb.New(a.CreatedAt),
	}
}

func convoyInfoToProto(c *ConvoyInfo) *pb.Convoy {
	return &pb.Convoy{
		Id:         c.ID,
		Name:       c.Name,
		Status:     c.Status,
		Formula:    c.Formula,
		ConfigJson: c.ConfigJSON,
		CreatedAt:  timestamppb.New(c.CreatedAt),
		UpdatedAt:  timestamppb.New(c.UpdatedAt),
	}
}

func mergeInfoToProto(m *MergeInfo) *pb.MergeRequest {
	return &pb.MergeRequest{
		Id:           m.ID,
		WorkItemId:   m.WorkItemID,
		RigId:        m.RigID,
		Branch:       m.Branch,
		TargetBranch: m.TargetBranch,
		Status:       m.Status,
		AgentId:      m.AgentID,
		QueuedAt:     timestamppb.New(m.QueuedAt),
	}
}

File: server/crates/server/src/routes/orchestration.rs
use axum::{
    Json, Router,
    extract::{Path, State},
    routing::{get, post},
};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::DeploymentImpl;

// ─── Types ───────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Town {
    pub id: String,
    pub name: String,
    pub owner: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Rig {
    pub id: String,
    pub town_id: String,
    pub name: String,
    pub repo_url: String,
    pub beads_prefix: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Agent {
    pub id: String,
    pub name: String,
    pub role: String,
    pub rig_id: String,
    pub status: String,
    pub runtime: String,
    pub last_activity_at: Option<String>,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Convoy {
    pub id: String,
    pub name: String,
    pub status: String,
    pub formula: String,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MergeRequestEntry {
    pub id: String,
    pub branch: String,
    pub target_branch: String,
    pub status: String,
    pub agent_id: String,
    pub pr_url: String,
    pub queued_at: String,
}

#[derive(Debug, Deserialize)]
pub struct CreateTownRequest {
    pub name: String,
    pub owner: String,
}

#[derive(Debug, Deserialize)]
pub struct CreateRigRequest {
    pub town_id: String,
    pub name: String,
    pub repo_url: String,
    pub beads_prefix: String,
}

#[derive(Debug, Deserialize)]
pub struct SpawnAgentRequest {
    pub name: String,
    pub role: String,
    pub rig_id: String,
    pub runtime: String,
}

#[derive(Debug, Deserialize)]
pub struct QueueMergeRequest {
    pub rig_id: String,
    pub branch: String,
    pub target_branch: String,
    pub agent_id: String,
}

// ─── Routes ──────────────────────────────────────────────────────────────────

pub fn router(_deployment: &DeploymentImpl) -> Router<DeploymentImpl> {
    Router::new()
        // Town
        .route("/orchestration/town", get(get_town))
        .route("/orchestration/town", post(create_town))
        // Rigs
        .route("/orchestration/rigs", get(list_rigs))
        .route("/orchestration/rigs", post(create_rig))
        // Agents
        .route("/orchestration/agents", get(list_agents))
        .route("/orchestration/agents/spawn", post(spawn_agent))
        .route("/orchestration/agents/{id}/kill", post(kill_agent))
        .route("/orchestration/agents/{id}", get(get_agent))
        // Convoys
        .route("/orchestration/convoys", get(list_convoys))
        .route("/orchestration/convoys/{id}", get(get_convoy))
        // Merge Queue
        .route("/orchestration/merge-queue", get(get_merge_queue))
        .route("/orchestration/merge-queue", post(queue_merge))
}

// ─── Handlers (stub implementations returning mock data) ─────────────────────
// These will be replaced with gRPC calls to the engine in Phase 1 integration.

async fn get_town(State(_deployment): State<DeploymentImpl>) -> Json<Town> {
    Json(Town {
        id: Uuid::new_v4().to_string(),
        name: "Default Town".to_string(),
        owner: "admin".to_string(),
    })
}

async fn create_town(
    State(_deployment): State<DeploymentImpl>,
    Json(req): Json<CreateTownRequest>,
) -> Json<Town> {
    Json(Town {
        id: Uuid::new_v4().to_string(),
        name: req.name,
        owner: req.owner,
    })
}

async fn list_rigs(State(_deployment): State<DeploymentImpl>) -> Json<Vec<Rig>> {
    Json(vec![Rig {
        id: Uuid::new_v4().to_string(),
        town_id: "default".to_string(),
        name: "vibetown".to_string(),
        repo_url: "https://github.com/edicorai/vibetown".to_string(),
        beads_prefix: "vt-".to_string(),
    }])
}

async fn create_rig(
    State(_deployment): State<DeploymentImpl>,
    Json(req): Json<CreateRigRequest>,
) -> Json<Rig> {
    Json(Rig {
        id: Uuid::new_v4().to_string(),
        town_id: req.town_id,
        name: req.name,
        repo_url: req.repo_url,
        beads_prefix: req.beads_prefix,
    })
}

async fn list_agents(State(_deployment): State<DeploymentImpl>) -> Json<Vec<Agent>> {
    Json(vec![])
}

async fn spawn_agent(
    State(_deployment): State<DeploymentImpl>,
    Json(req): Json<SpawnAgentRequest>,
) -> Json<Agent> {
    Json(Agent {
        id: Uuid::new_v4().to_string(),
        name: req.name,
        role: req.role,
        rig_id: req.rig_id,
        status: "idle".to_string(),
        runtime: req.runtime,
        last_activity_at: None,
        created_at: chrono::Utc::now().to_rfc3339(),
    })
}

async fn kill_agent(
    State(_deployment): State<DeploymentImpl>,
    Path(id): Path<String>,
) -> Json<serde_json::Value> {
    Json(serde_json::json!({"id": id, "status": "killed"}))
}

async fn get_agent(
    State(_deployment): State<DeploymentImpl>,
    Path(id): Path<String>,
) -> Json<Agent> {
    Json(Agent {
        id,
        name: "agent-1".to_string(),
        role: "polecat".to_string(),
        rig_id: "default".to_string(),
        status: "idle".to_string(),
        runtime: "claude".to_string(),
        last_activity_at: None,
        created_at: chrono::Utc::now().to_rfc3339(),
    })
}

async fn list_convoys(State(_deployment): State<DeploymentImpl>) -> Json<Vec<Convoy>> {
    Json(vec![])
}

async fn get_convoy(
    State(_deployment): State<DeploymentImpl>,
    Path(id): Path<String>,
) -> Json<Convoy> {
    Json(Convoy {
        id,
        name: "convoy-1".to_string(),
        status: "active".to_string(),
        formula: "".to_string(),
        created_at: chrono::Utc::now().to_rfc3339(),
    })
}

async fn get_merge_queue(
    State(_deployment): State<DeploymentImpl>,
) -> Json<Vec<MergeRequestEntry>> {
    Json(vec![])
}

async fn queue_merge(
    State(_deployment): State<DeploymentImpl>,
    Json(req): Json<QueueMergeRequest>,
) -> Json<MergeRequestEntry> {
    Json(MergeRequestEntry {
        id: Uuid::new_v4().to_string(),
        branch: req.branch,
        target_branch: req.target_branch,
        status: "pending".to_string(),
        agent_id: req.agent_id,
        pr_url: "".to_string(),
        queued_at: chrono::Utc::now().to_rfc3339(),
    })
}
